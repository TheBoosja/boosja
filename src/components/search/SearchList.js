import React, { Component } from 'react';
import { connect } from 'react-redux';
import { search, emptyResults } from 'actions/search';
import SearchListItem from './SearchListItem';
import { SearchResult, ResultList, TertiaryHeading } from 'components/styled';

export class SearchList extends Component {
	constructor(props) {
		super(props);

		this.handleQuery = this.handleQuery.bind(this);
	}

	handleQuery(query, page = 1) {
		this.props.search(query, page);
	}

	componentWillMount() {
		const { query, page } = this.props.match.params;
		this.handleQuery(query, page);
	}

	componentWillUpdate({ match: { params: { query } } }) {
		const oldQuery = this.props.match.params.query;

		if (query !== oldQuery) {
			this.props.emptyResults();
			this.handleQuery(query);
		}
	}

	renderResults(results) {
		return results.map((res, i) => (
			<li key={i}>
				<SearchListItem show={res} />
			</li>
		));
	}

	render() {
		const { results, total_results, total_pages, match: { params: { page } } } = this.props;

		if (!results) {
			return <span>Loading...</span>;
		}

		return (
			<SearchResult>
				<TertiaryHeading>Results: {total_results}</TertiaryHeading>

				<ResultList>{this.renderResults(results)}</ResultList>
			</SearchResult>
		);
	}
}

function mapStateToProps({ search: { results, total_results, page, total_pages } }) {
	return {
		results,
		total_results,
		page,
		total_pages
	};
}

export default connect(mapStateToProps, { search, emptyResults })(SearchList);
