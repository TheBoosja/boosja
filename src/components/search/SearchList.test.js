import React from 'react';
import { shallow } from 'enzyme';
import { SearchList } from './SearchList';

describe('Search List', () => {
	const { query, page } = { query: 'lady', page: 2 };
	const results = [{ name: 'test1' }, { name: 'test2' }];

	const mockSearch = jest.fn();
	const mockEmptyResults = jest.fn();
	const props = {
		match: { params: { query, page } },
		results,
		search: mockSearch,
		emptyResults: mockEmptyResults
	};
	const searchList = shallow(<SearchList {...props} />);

	it('renders properly', () => {
		expect(searchList).toMatchSnapshot();
	});

	it('renders a list of SearchListItems', () => {
		expect(searchList.find('SearchListItem').length).toEqual(results.length);
	});

	describe('when mounting', () => {
		searchList.instance().componentWillMount();

		it('calls the search method with the query and page', () => {
			expect(mockSearch).toHaveBeenCalledWith(query, page);
		});
	});

	describe('when updating', () => {
		let newMockSearch;

		beforeEach(() => {
			newMockSearch = jest.fn();
			searchList.setProps({ search: newMockSearch });
		});

		it('calls the search method when the query is new', () => {
			const newQuery = 'heroes';
			searchList.instance().componentWillUpdate({ match: { params: { query: newQuery } } });

			expect(newMockSearch).toHaveBeenCalledWith(newQuery, 1);
		});

		it('does not call the search method when the query is the same', () => {
			searchList.instance().componentWillUpdate({ match: { params: { query } } });

			expect(newMockSearch).not.toHaveBeenCalled();
		});
	});

	describe('when results is empty', () => {
		const emptyProps = {
			match: { params: { query, page } },
			results: undefined,
			search: mockSearch,
			emptyResults: mockEmptyResults
		};
		const emptySearchList = shallow(<SearchList {...emptyProps} />);

		it('renders loading indicator', () => {
			expect(emptySearchList.find('span').text()).toEqual('Loading...');
		});
	});
});
