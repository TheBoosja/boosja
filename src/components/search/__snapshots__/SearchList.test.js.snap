// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search List renders properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SearchList
    emptyResults={[Function]}
    match={
        Object {
            "params": Object {
              "page": 2,
              "query": "lady",
            },
          }
    }
    results={
        Array [
            Object {
              "name": "test1",
            },
            Object {
              "name": "test2",
            },
          ]
    }
    search={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <TertiaryHeading>
          Results: 
</TertiaryHeading>,
        <ResultList>
          <li>
                    <SearchListItem
                              show={
                                        Object {
                                                  "name": "test1",
                                                }
                              }
                    />
          </li>
          <li>
                    <SearchListItem
                              show={
                                        Object {
                                                  "name": "test2",
                                                }
                              }
                    />
          </li>
</ResultList>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            "Results: ",
            undefined,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Results: ",
          undefined,
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <li>
              <SearchListItem
                            show={
                                          Object {
                                                        "name": "test1",
                                                      }
                            }
              />
</li>,
            <li>
              <SearchListItem
                            show={
                                          Object {
                                                        "name": "test2",
                                                      }
                            }
              />
</li>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "host",
            "props": Object {
              "children": <SearchListItem
                show={
                                Object {
                                                "name": "test1",
                                              }
                }
/>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "show": Object {
                  "name": "test1",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "li",
          },
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "host",
            "props": Object {
              "children": <SearchListItem
                show={
                                Object {
                                                "name": "test2",
                                              }
                }
/>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "show": Object {
                  "name": "test2",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "li",
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <TertiaryHeading>
            Results: 
</TertiaryHeading>,
          <ResultList>
            <li>
                        <SearchListItem
                                    show={
                                                Object {
                                                            "name": "test1",
                                                          }
                                    }
                        />
            </li>
            <li>
                        <SearchListItem
                                    show={
                                                Object {
                                                            "name": "test2",
                                                          }
                                    }
                        />
            </li>
</ResultList>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              "Results: ",
              undefined,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Results: ",
            undefined,
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <li>
                <SearchListItem
                                show={
                                                Object {
                                                                "name": "test1",
                                                              }
                                }
                />
</li>,
              <li>
                <SearchListItem
                                show={
                                                Object {
                                                                "name": "test2",
                                                              }
                                }
                />
</li>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "0",
              "nodeType": "host",
              "props": Object {
                "children": <SearchListItem
                  show={
                                    Object {
                                                      "name": "test1",
                                                    }
                  }
/>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "show": Object {
                    "name": "test1",
                  },
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "li",
            },
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "host",
              "props": Object {
                "children": <SearchListItem
                  show={
                                    Object {
                                                      "name": "test2",
                                                    }
                  }
/>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "show": Object {
                    "name": "test2",
                  },
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "li",
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
